File: 2024/day02-1.ts
function solution(content: string) {
  return content
    .split("\n")
    .map((line) => line.split(" ").map(Number))
    .filter(isSafe).length;
}

function isSafe(level: number[]) {
  return isWithinRange(level) && (isDecreasing(level) || isIncreasing(level));
}

function isIncreasing(level: number[]) {
  return level.every((item, index, array) =>
    index < array.length - 1 ? item < array[index + 1] : true
  );
}

function isDecreasing(level: number[]) {
  return level.every((item, index, array) =>
    index < array.length - 1 ? item > array[index + 1] : true
  );
}

function isWithinRange(level: number[]) {
  return level.every((item, index, array) =>
    index < array.length - 1
      ? 1 <= Math.abs(item - array[index + 1]) &&
        Math.abs(item - array[index + 1]) <= 3
      : true
  );
}

Deno.readTextFile("day02.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2024/day02-1.hs
import System.IO

solution :: String -> Int
solution = length . filter id . map isSafe . map (map read) . map words . lines

isSafe :: [Int] -> Bool
isSafe xs = (isWithinRange xs) && ((isIncreasing xs) || (isDecreasing xs))

isIncreasing :: [Int] -> Bool
isIncreasing xs = all (\(a, b) -> b > a) (zip xs (tail xs))

isDecreasing :: [Int] -> Bool
isDecreasing xs = all (\(a, b) -> b < a) (zip xs (tail xs))

isWithinRange :: [Int] -> Bool
isWithinRange xs = all (\(a, b) -> 1<=abs(a-b) && abs(a-b)<=3) (zip xs (tail xs))

main :: IO ()
main = do
    handle <- openFile "day02.dat" ReadMode
    content <- hGetContents handle
    print . solution $ content